stan/math/opencl/triangular.hpp

```diff
@@ -0,0 +1,61 @@
+#ifndef STAN_MATH_OPENCL_TRIANGULAR_HPP
+#define STAN_MATH_OPENCL_TRIANGULAR_HPP
+#ifdef STAN_OPENCL
+
+#include <type_traits>
+
+namespace stan {
+namespace math {
+enum class TriangularViewCL { Diagonal = 0, Lower = 1, Upper = 2, Entire = 3 };
+
+inline TriangularViewCL operator|(TriangularViewCL a, TriangularViewCL b) {
+  typedef typename std::underlying_type<TriangularViewCL>::type underlying;
+  return static_cast<TriangularViewCL>(static_cast<underlying>(a)
+                                       | static_cast<underlying>(b));
+}
+
+inline TriangularViewCL operator|=(TriangularViewCL& a, TriangularViewCL b) {
+  a = a | b;
+  return a;
+}
+
+inline TriangularViewCL operator&(TriangularViewCL a, TriangularViewCL b) {
+  typedef typename std::underlying_type<TriangularViewCL>::type underlying;
+  return static_cast<TriangularViewCL>(static_cast<underlying>(a)
+                                       & static_cast<underlying>(b));
+}
+
+inline TriangularViewCL operator&=(TriangularViewCL& a, TriangularViewCL b) {
+  a = a & b;
+  return a;
+}
+
+inline TriangularViewCL transpose(TriangularViewCL a) {
+  switch (a) {
```

Comment:
I'd rather have if statements here

Reply:
why?

Reply:
These are just suggestions---@SteveBronder is still in charge of the review.

1.  Shouldn't this be a metaprogram as it's only returning a type?  Something like:

```
template <typename T>
struct tri_transpose {
};

template <>
struct tri_tranpose<Lower> {
  typedef Upper type;
};
...
```

and usage would be `typename tri_tranpose<TrinagularViewCl::Lower>::type`.

2.  I think we should extend this to the other view type pairs: `Upper, Lower`, `StrictlyUpper, StrictlyLower`, and `UnitUpper, UnitLower`.

3.  These programs need doxygen doc.

Reply:
1. No. These funtions are called at runtime, so they can not be reworked into templates. 

2. Adding each of those would add a bit to complexity of manipulations of triangular views. Right now I see no need for them. They can be added at later time if some new functionality requires them.

3. Ok, I will add doxy.

Reply:
For (1), what's the use case?  Specifically, when is the value of `a` not known at compile time?  I don't see how you could take one of these enums and plug it back in.

For (2), if you don't want to complete the function naturally, I would suggest throwing an exception rather than returning a clearly wrong value by default.  

For (3), thanks.  I think the problem posed by (2) will be clearer when you try to write doc.

Reply:
1. The way things are implemented now verywhere. The most obvious case where making it static would be very hard is `.sub_block()`. The view of the result depends on view of the inputs and arguments - to which part of a matrix it is copying.

2. Now we don't even have enum values for `StrictlyUpper`, `StrictlyLower`, `UnitUpper`, `UnitLower`.

Reply:
> why?

It's mostly that we use if statements everywhere else. Though they are more efficient I personally don't like using switches because of the fall-through gotcha.

Bob for context, doing this at compile time is possible for pretty much every function besides `sub_block`. I would bet there is some weird algebra we could do at compile time to figure out the triangularity when inserting one matrix into another, but I thought about it for about 10 minutes and it started hurting my brain.

So instead of doing a mix of static and dynamic we chose to go dynamic. I'm still not against doing as much at compile time as we can